ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -688,7 +688,7 @@ Operator genOPTree(ASTNode ast, PlannerContext plannerCtx) throws SemanticExcept
           this.ctx.setCboInfo(cboMsg);
 
           // Determine if we should re-throw the exception OR if we try to mark the query to retry as non-CBO.
-          if (fallbackStrategy.isFatal(e)) {
+          if (fallbackStrategy.isFatal(e) || HiveCalciteUtil.requiresCBO(ast)) {
```

Comment:
The additional disjunctive clause changes a bit the behavior of the fallback strategies notably when the `ALWAYS/CONSERVATIVE` option is set. 

Nevertheless, it makes sense to avoid recompiling a query when we know that it will fail so the check fits here. Let's just check if we really need a new method and a tree traversal as I mentioned in the other comment.